### PR TITLE
add exit_code log in app_manager

### PR DESCRIPTION
**Pull request by knorth55**
add loginfo and logerr for `exit_code` with app launch process.

### CODE REVIEW

File: src/app_manager/app_manager.py [hunk 1]
```diff
@@ -250,6 +250,14 @@ def handle_start_app(self, req):
...
         try:
             self._launch.shutdown()
+            if len(self._launch.pm.dead_list) > 0:
+                exit_code = self._launch.pm.dead_list[0].exit_code
+                if exit_code > 0:
+                    rospy.logerr(
+                        "App stopped with exit code: {}".format(exit_code))
+                else:
+                    rospy.loginfo(
+                        "App stopped with exit code: {}".format(exit_code))
```
Comment (v4hn) [inline src/app_manager/app_manager.py]: Could you leave out the info for `exit_code == 0`?
Why would you log successful termination like that?

Comment (knorth55) [reply src/app_manager/app_manager.py]: There are possibly three cases.
- `len(pm.dead_list) == 0`: success
- `len(pm.dead_list) > 0`: something wrong
  - `pm.dead_list[0].exit_code == 0`: main process is OK
  - `pm.dead_list[0].exit_code > 1`: main process died

https://github.com/ros/ros_comm/blob/noetic-devel/tools/roslaunch/src/roslaunch/pmon.py#L558-L624
According lines here, if `pm` `shutdown` correctly, `len(pm.dead_list)` is `0`.
Therefore, if `len(pm.dead_list)` is not `0`, something went wrong with process.
Therefore, I add `rospy.loginfo` for `exit_code==0`.
But I think, for both case, we can use `rospy.logerr`.

Comment (v4hn) [reply src/app_manager/app_manager.py]: If that's the case, please simplify and remove the condition. Please also change the message to imply the app should not have been in the list anymore.

Comment (knorth55) [reply src/app_manager/app_manager.py]: > Please also change the message to imply the app should not have been in the list anymore.

I'm sorry, but I don't get your point.
What do you mean?
Do i need to add `rospy.loginfo` for the case `len(pm.dead_list) == 0`?

Comment (knorth55) [reply src/app_manager/app_manager.py]: > If that's the case, please simplify and remove the condition. 

I did this change. Thank you for your review!

Comment (v4hn) [reply src/app_manager/app_manager.py]: Sorry, but staring at the code you referenced, I don't see why `dead_list` *must* be empty on shutdown.

[this line](https://github.com/ros/ros_comm/blob/noetic-devel/tools/roslaunch/src/roslaunch/pmon.py#L574) might well add elements to `dead` before [this line](https://github.com/ros/ros_comm/blob/noetic-devel/tools/roslaunch/src/roslaunch/pmon.py#L586) breaks the inner polling (for) loop and the [next loop](https://github.com/ros/ros_comm/blob/noetic-devel/tools/roslaunch/src/roslaunch/pmon.py#L599) adds those elements to `dead_list`.
Am I mistaken?

Comment (knorth55) [reply src/app_manager/app_manager.py]: I thinks there are three cases of app stop.
### required process ended
the  main process have `required` tag, so these lines are used https://github.com/ros/ros_comm/blob/noetic-devel/tools/roslaunch/src/roslaunch/pmon.py#L566-L568
in this case, `pm.dead_list` is empty.

### run service call `stop_app`
when you run `stop_app` service, `pm.is_shutdown` flag is raised, and goes into this line https://github.com/ros/ros_comm/blob/noetic-devel/tools/roslaunch/src/roslaunch/pmon.py#L586 directly.
in this case, `pm.dead_list` is empty, too.

### process died with some exit_code
we can catch exit_code in `pm.dead_list`
in this case, `pm.dead_list` is not empty.

Comment (v4hn) [reply src/app_manager/app_manager.py]: https://github.com/ros/ros_comm/blob/noetic-devel/tools/roslaunch/src/roslaunch/pmon.py#L586 does not break the full `while` loop, only the internal `for` loop. Doesn't that mean all elements in `dead` will still end up in `dead_list`?

Comment (knorth55) [reply src/app_manager/app_manager.py]: If app is running correctly, required process stops earlier than other process when it is done.
so these lines will run https://github.com/ros/ros_comm/blob/noetic-devel/tools/roslaunch/src/roslaunch/pmon.py#L566-L568
in these lines `pm.is_shutdown = True` is set, so https://github.com/ros/ros_comm/blob/noetic-devel/tools/roslaunch/src/roslaunch/pmon.py#L549-L550 breaks the while loop in next step

Comment (v4hn) [reply src/app_manager/app_manager.py]: Ok, thanks for bearing with me on this.

**Review by v4hn — APPROVED** [review]

File: src/app_manager/app_manager.py (hunk 1 above)
Comment (knorth55) [reply src/app_manager/app_manager.py]: no problem.
you can try with simple apps. I made test repository for this.
https://github.com/knorth55/test_app_manager

```
roslaunch app_manager app_manager.launch
rosservice call /robot/start_app "name: 'test_app_manager/test_app'"
```

**Pull request merged**